plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,247 @@
+findSigLocOfStringDecl decls declName = do
+    -- can we simplify this logic? Just want to make sure ghcSig is a Just value
+    (RealSrcSpan rss _, Just ghcSig) <- listToMaybe [ (locA ss, sigToText ts)
+                                        | L ss (SigD _ ts@(TypeSig _ idsSig _)) <- decls,
+                                        any ((==) declName . occNameString . occName . unLoc) idsSig
+                                      ]
+    pure (rss, ghcSig)
+
+-- | Pretty Print the Type Signature (to validate GHC Error Message)
+sigToText :: Sig GhcPs -> Maybe Text
+sigToText = \case
+  ts@TypeSig {} -> stripSignature (T.pack $ showSDocUnsafe $ ppr ts)
+  _             -> Nothing
+
+stripSignature :: Text -> Maybe Text
+-- for whatever reason incoming signatures MAY have new lines after "::" or "=>"
+stripSignature sig = case T.filter (/= '\n') sig =~ sigRegex :: (Text, Text, Text, [Text]) of
+                        -- No constraints (Monad m =>)
+                         (_, _, _, [sig'])    -> Just $ T.strip sig'
+                        -- Ignore constraints (Monad m =>)
+                         (_, _, _, [_, sig']) -> Just $ T.strip sig'
+                         _                    -> Nothing
+    where
+        -- we want to test everthing after the constraints (GHC never gives us the constraint in the expected signature)
+        sigRegex = ".* :: (.*=>)?(.*)" :: Text
+
+
+changeSigToCodeAction :: Uri -> ChangeSignature -> Command |? CodeAction
+changeSigToCodeAction uri ChangeSignature{..} = InR CodeAction { _title       = mkChangeSigTitle declName actualType
+                                                               , _kind        = Just CodeActionQuickFix
+                                                               , _diagnostics = Just $ List [diagnostic]
+                                                               , _isPreferred = Nothing
+                                                               , _disabled    = Nothing
+                                                               , _edit        = Just $ mkChangeSigEdit uri declSrcSpan (mkNewSignature declName actualType)
+                                                               , _command     = Nothing
+                                                               , _xdata       = Nothing
+                                                               }
+
+mkChangeSigTitle :: Text -> Text -> Text
+mkChangeSigTitle declName actualType = "change signature for ‘" <> declName <> "’ to: " <> actualType
+
+mkChangeSigEdit :: Uri -> RealSrcSpan -> Text -> WorkspaceEdit
+mkChangeSigEdit uri ss replacement =
+        let txtEdit = TextEdit (realSrcSpanToRange ss) replacement
+            changes = Just $ Map.singleton uri (List [txtEdit])
+        in WorkspaceEdit changes Nothing Nothing
+
+mkNewSignature :: Text -> Text -> Text
+mkNewSignature declName actualType = declName <> " :: " <> actualType
```

Comment:
Perhaps stupid question, but should we be using exactprint somehow to generate stuff like this?

Reply:
For example: what happens if there are pragmas or something attached to the declaration? will they get clobbered? If they're included in the decl span they might do.

Reply:
I'll look into it

Reply:
I did some testing and I don't see any issues with Language Pragmas near signatures. I'm also not really familiar with `exactprint`. Considering we have to parse the GHC message to get the signature it seems easier to just to build it up ourselves?

Reply:
I was thinking more pragmas that might get attached to the declaration like `{-# INLINE #-}`.

I'm also not really familiar with it: I just note that if the chunk of AST you're replacing includes stuff that isn't just the signature, you'll clobber it. I think this is the sort of thing exactprint solves, but I also haven't used it in anger, perhaps @pepeiborra can weigh in on whether this is something to worry about?

Reply:
Using exactprint to edit the AST is always the principled thing to do, we already have helpers to edit existing type signatures:

https://github.com/haskell/haskell-language-server/blob/4b8f9d53bed3e775ccd120d9a2471ae374a88db3/ghcide/src/Development/IDE/Plugin/CodeAction/ExactPrint.hs#L16-L18

But admittedly, exactprint is really hard to use, and the api changes in 9.2 make it even harder!

Reply:
> I was thinking more pragmas that might get attached to the declaration like `{-# INLINE #-}`.

Yes sorry, not sure why I said Language Pragma. Those are the ones I meant!